# Практическая работа №2: Исследование алгоритмов формирования аддитивных цепочек

Выполнил студент гр. 1304 Басыров Владимир. Вариант №30.

## Цель работы
Формирование представления о аддитивных цепочках, выработать уме-
ние составлять и применять алгоритмы для нахождения минимальных
аддитивных цепочек для заданного числа, привить навык использования
систем компьютерной математики для реализации алгоритмов.


## Основные теоретические положения

#### Аддитивные цепочки

Аддитивная цепочка для некоторого числа $n\in \mathbb{N}$ - это последовательность натуральных чисел $\{a_i\}_{i=0}^m$, начинающаяся с единицы, в которой каждый последующий элемент является суммой двух любых предшествующих элементов. Эта последовательность удовлетворяет свойствам:
1. $a_0 = 1$;
2. $\forall i > 0: a_i = a_j + a_k$, где $j, k < i$.

Длина аддитивной цепочки $l(n)=|\{a_i\}_{i=0}^m|$, где $a_m = n$.

$a_i=a_j+a_k$ - шаг аддитивной цепочки, $i \in \{1,2 \dots m\}$, $0\leqslant k\leqslant j < i$

Типы шагов в аддитивной цепочке:
1. Удвоение:  $i - 1 = k = j$;
2. Звездный шаг: $j = i - 1$, $k \in \{0, \dots, i-1\}$;
3. Малый шаг: $\lambda(a_i)=\lambda(a_{i-1})$

Звездная цепочка - это аддитивная цепочка, в которая состоит только из звездных шагов.

#### Алгоритм Яо

Алгоритм Яо - это алгоритм для поиска аддитивной цепочки для некоторого числа $n \in \mathbb{N}$. Cначала фиксируются переменные $n,k \geq 2$, после чего $n$ представляется в виде $\sum_{i=0} a_j \cdot 2^{ik}$, где $a_j \neq 0$. 

Далее вводится функция $d(z)=\sum_{\forall i: a_i=z} 2^{ik}$.

В начало аддитивной цепочки Яо добавляются степени двойки: $Y_k(n): 1,2,4,\dots ,2^{\lambda(n)}$, после чего вычисляются все значения $d(z)$, где $z \in \{1,2,3, \dots ,2^k-1 \}$ и $d(z) \neq 0$.

В таком случае число $n=\sum_{z=1}^{2^k-1}z \cdot d(z)$.

#### Алгоритм дробления вектора индексов 

Алгоритм дробления вектора индексов позволяет найти минимальную звездную цепочку для некоторого числа $n \in \mathbb{N}$. 

Рассмотрим вектор индексов $\{r_i\}_{i=1}^q \cup {\{{\rho}_j \}}_{j=q+1}^m$, где ${\rho}_j= \{x: 1 \leq x \leq j \}$, ${\{r_i\}}_{i=1}^q$ - фиксированная часть, ${\{{\rho}_j\}}_{j=q+1}^m$ - изменяющаяся часть.

$a_{max}$ достигается при векторе индексов ${\{r_i\}}_{i=1}^{q} \cup \{q+1,q+2,\dots,m\}$.

$a_{min}$ достигается при векторе индексов ${\{r_i\}}_{i=1}^{q}\cup\{1,1,\dots,1\}$.

$a_{max} = a_{q+1} \cdot {2}^{m-q}$

$a_{min} = a_{q+1}+m-q$

Алгоритм:
1. Во внешнем цикле рассматриваем аддитивные цепочки длины $m$ от значения $\bar{l}(n)=\lceil log_2(n) \rceil$ до $\underline{l}(n)=\lambda(n)+\nu(n)-1$, на каждой итерации выбираем $q$ ($1 \leq q \leq m-1$);
2. Далее перебираем все возможные фиксированные части вектора индексов $\{r_i\}_{i=1}^q$ ($q!$ вариантов), для каждой строим соответствующую ей звездную цепочку, находим $a_{max}$ и ${a}_{min}$, после чего:
 1. Если $n \notin [a_{min},a_{max}]$, то переходим к следующему набору $\{r_i\}_{i=1}^q$;
 2. Если $n\in [a_{min},a_{max}]$, то перебираем все возможные изменяющиеся части вектора индексов ${\{{\rho}_j\}}_{j=q+1}^m$ и находим $a_m$:
  1. Если $a_m=n$, то цепочка найдена;
  2. Если все возможные изменяющиеся части вектора индексов ${\{{\rho}_j\}}_{j=q+1}^m$ исчерпаны, то переходим к следующему набору $\{r_i\}_{i=1}^q$;
3. Если все наборы вектора индексов длины $m$ исчерпаны, то увеличиваем $m$ на 1.


#### Гипотеза Шольца-Брауэра
Пусть $l^*(n)$ - длина некоторой звёздной цепочки.

Тогда для любого $n \in \mathbb{N}$ верно: $l^*(2^n-1)\leq l^*(n)+n-1$. Равенство справедливо для любого $n<=64$

## Постановка задачи



## Выполнение работы

### Алгоритм Яо

Для реализации алгоритма Яо воспользуемся рядом вспомогатльных функций. Так, была реализована функция binary_method, который вычисляет коэффиценты для получения выражения $z*d(z)$.

In [3]:
def binary_method(n):
    binary_n=bin(n)[3:]
    res=[1]
    number=1
    for i in range(len(binary_n)):
        number+=number
        res.append(number)
        if int(binary_n[i]):
            number+=1
            res.append(number)
    return res
print(binary_method(15))

[1, 2, 3, 6, 7, 14, 15]


Для вычисления представления n-арной записи числа была создана функция dz_array_mod, которая принимает число и систему, в которой нужно представить это число. 

In [21]:
def dz_array_mod(n,k):
    k_system=[]
    number=n
    while number>=k:
        k_system.append(number%k)
        number//=k
    k_system.append(number)
    return k_system

Функция вычисления $d(z)$,по заданному $z$, системе $k$ , и представлению числа в системе счислени

In [22]:
def dz(z,k,array_mod):
    dz_z=0
    res=[]
    mul=1
    for i in range(len(array_mod)):
        if z==array_mod[i]:
            dz_z+=mul
            res.append(dz_z)
        mul*=k
    return dz_z,res

В совокупности с предыдущими функциями представлен алгоритм Яо.

In [ ]:
def algoritm_YAO(n,k):
    array_mod=dz_array_mod(n,k)
    result=[2**i for i in range(int(log(n,2))+1)]
    dz_z_array=[]
    for z in range(1,k):
        dz_z,res_dz_z=dz(z,k,array_mod)
        if dz_z:
            dz_z_array.append(z*dz_z)
            result.extend(res_dz_z)
            binary_result=binary_method(z)[1:]
            for i in binary_result:
                result.append(dz_z*i)
    if len(dz_z_array):            
        sum_dz=dz_z_array[0]
        for i in dz_z_array[1:]:
            sum_dz+=i
            result.append(sum_dz)
    result.sort()
    i=0
    len_result=len(result)
    while i<len_result-1:
        while i<len_result-1 and result[i]==result[i+1]:
            result.pop(i)
            len_result-=1
        i+=1
    return result
dict_test={
    13 : [1,2,3,6,12,13],
    232: [1,2,3,6,7,14,28,29,58,116,232],
    511: [1,2,3,6,12,15,30,60,120,240,255,510,511],
    723: [1,2,3,5,10,20,40,45,90,180,360,720,723],
    1024: [1,2,4,8,16,32,64,128,256,512,1024],
    1345: [1,2,4,5,10,20,21,42,84,168,336,672,1344,1345]
}
array_n=[13,232,511,723,1024,1345]
array_k=[2,3,4,5,6]
print("|n|k|Длина минимальной цепочки Яо|Минимальная Цепочка Яо|Длина минимальной цепочки|Минимальная цепочка|")
for n in array_n:
    for k in array_k:
        result=algoritm_YAO(n,k)
        print('|',n,'|',k,'|',len(result),'|',result,'|',len(dict_test[n]),'|',dict_test[n],'|')

Была составлена талица варьирования параметра k и n соотвественно

|n|k|Длина минимальной цепочки Яо|Минимальная Цепочка Яо|Длина минимальной цепочки|Минимальная цепочка|
|-|-|----------------------------|----------------------|-------------------------|-------------------|
| 13 | 2 | 6 | [1, 2, 4, 5, 8, 13] | 6 | [1, 2, 3, 6, 12, 13] |
| 13 | 3 | 6 | [1, 2, 4, 5, 8, 13] | 6 | [1, 2, 3, 6, 12, 13] |
| 13 | 4 | 6 | [1, 2, 4, 8, 12, 13] | 6 | [1, 2, 3, 6, 12, 13] |
| 13 | 5 | 8 | [1, 2, 3, 4, 5, 8, 10, 13] | 6 | [1, 2, 3, 6, 12, 13] |
| 13 | 6 | 7 | [1, 2, 4, 6, 8, 12, 13] | 6 | [1, 2, 3, 6, 12, 13] |
| 232 | 2 | 11 | [1, 2, 4, 8, 16, 32, 40, 64, 104, 128, 232] | 11 | [1, 2, 3, 6, 7, 14, 28, 29, 58, 116, 232] |
| 232 | 3 | 14 | [1, 2, 3, 4, 8, 10, 16, 30, 32, 64, 111, 128, 222, 232] | 11 | [1, 2, 3, 6, 7, 14, 28, 29, 58, 116, 232] |
| 232 | 4 | 12 | [1, 2, 4, 8, 16, 20, 32, 40, 64, 128, 192, 232] | 11 | [1, 2, 3, 6, 7, 14, 28, 29, 58, 116, 232] |
| 232 | 5 | 15 | [1, 2, 4, 5, 8, 16, 25, 32, 50, 64, 100, 128, 130, 132, 232] | 11 | [1, 2, 3, 6, 7, 14, 28, 29, 58, 116, 232] |
| 232 | 6 | 13 | [1, 2, 4, 6, 8, 12, 16, 32, 64, 128, 216, 228, 232] | 11 | [1, 2, 3, 6, 7, 14, 28, 29, 58, 116, 232] |
| 511 | 2 | 17 | [1, 2, 3, 4, 7, 8, 15, 16, 31, 32, 63, 64, 127, 128, 255, 256, 511] | 13 | [1, 2, 3, 6, 12, 15, 30, 60, 120, 240, 255, 510, 511] |
| 511 | 3 | 14 | [1, 2, 3, 4, 8, 12, 16, 32, 64, 128, 255, 256, 510, 511] | 13 | [1, 2, 3, 6, 12, 15, 30, 60, 120, 240, 255, 510, 511] |
| 511 | 4 | 15 | [1, 2, 4, 5, 8, 16, 21, 32, 64, 85, 128, 170, 255, 256, 511] | 13 | [1, 2, 3, 6, 12, 15, 30, 60, 120, 240, 255, 510, 511] |
| 511 | 5 | 16 | [1, 2, 4, 5, 8, 10, 11, 16, 32, 64, 125, 128, 250, 256, 500, 511] | 13 | [1, 2, 3, 6, 12, 15, 30, 60, 120, 240, 255, 510, 511] |
| 511 | 6 | 14 | [1, 2, 4, 7, 8, 16, 32, 36, 64, 128, 252, 256, 504, 511] | 13 | [1, 2, 3, 6, 12, 15, 30, 60, 120, 240, 255, 510, 511] |
| 723 | 2 | 15 | [1, 2, 3, 4, 8, 16, 19, 32, 64, 83, 128, 211, 256, 512, 723] | 13 | [1, 2, 3, 5, 10, 20, 40, 45, 90, 180, 360, 720, 723] |
| 723 | 3 | 17 | [1, 2, 3, 4, 8, 9, 16, 32, 36, 64, 117, 128, 256, 360, 512, 720, 723] | 13 | [1, 2, 3, 5, 10, 20, 40, 45, 90, 180, 360, 720, 723] |
| 723 | 4 | 15 | [1, 2, 4, 8, 16, 32, 64, 65, 128, 130, 195, 256, 512, 528, 723] | 13 | [1, 2, 3, 5, 10, 20, 40, 45, 90, 180, 360, 720, 723] |
| 723 | 5 | 19 | [1, 2, 4, 5, 8, 10, 16, 20, 26, 32, 52, 64, 78, 128, 256, 512, 625, 703, 723] | 13 | [1, 2, 3, 5, 10, 20, 40, 45, 90, 180, 360, 720, 723] |
| 723 | 6 | 16 | [1, 2, 4, 8, 16, 32, 36, 64, 72, 128, 217, 256, 434, 512, 651, 723] | 13 | [1, 2, 3, 5, 10, 20, 40, 45, 90, 180, 360, 720, 723] |
| 1024 | 2 | 11 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024] | 11 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024] |
| 1024 | 3 | 17 | [1, 2, 3, 4, 8, 12, 16, 24, 28, 32, 64, 128, 256, 271, 512, 1000, 1024] | 11 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024] |
| 1024 | 4 | 11 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024] | 11 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024] |
| 1024 | 5 | 19 | [1, 2, 4, 6, 8, 12, 16, 24, 32, 64, 125, 128, 250, 256, 375, 512, 625, 1000, 1024] | 11 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024] |
| 1024 | 6 | 17 | [1, 2, 4, 6, 8, 12, 16, 32, 37, 64, 128, 253, 256, 506, 512, 1012, 1024] | 11 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024] |
| 1345 | 2 | 14 | [1, 2, 4, 8, 16, 32, 64, 65, 128, 256, 321, 512, 1024, 1345] | 14 | [1, 2, 4, 5, 10, 20, 21, 42, 84, 168, 336, 672, 1344, 1345] |
| 1345 | 3 | 18 | [1, 2, 4, 8, 9, 16, 31, 32, 64, 112, 128, 252, 256, 504, 512, 841, 1024, 1345] | 14 | [1, 2, 4, 5, 10, 20, 21, 42, 84, 168, 336, 672, 1344, 1345] |
| 1345 | 4 | 14 | [1, 2, 4, 8, 16, 32, 64, 65, 128, 256, 321, 512, 1024, 1345] | 14 | [1, 2, 4, 5, 10, 20, 21, 42, 84, 168, 336, 672, 1344, 1345] |
| 1345 | 5 | 21 | [1, 2, 4, 5, 8, 10, 16, 20, 25, 32, 50, 64, 75, 128, 256, 512, 625, 1024, 1250, 1325, 1345] | 14 | [1, 2, 4, 5, 10, 20, 21, 42, 84, 168, 336, 672, 1344, 1345] |
| 1345 | 6 | 16 | [1, 2, 4, 6, 8, 12, 16, 32, 37, 64, 128, 256, 512, 1024, 1333, 1345] | 14 | [1, 2, 4, 5, 10, 20, 21, 42, 84, 168, 336, 672, 1344, 1345] |


### Вывод.
Алгоритм Яо дает цепочку за минимально возможное время, однако это цепочка не всегда может быть минимальной, ее минимальность зависит от правильного выбора элемента k.

### Алгоритм дробления вектора

Для вычисления нижней и верхней оценки были разработаны слудующие функции.

In [8]:
def lower_estimate(n):
    return int(log(n,2))
def high_estimate(n):
    return lower_estimate(n)+list(bin(n)).count('1')-1

Для корректной работы алгоритма также был разработан ряд вспомогательных функций. Так функция next_variant, по задданому вектору, размеру этого вектора и тому, является ли этот вектор $p\_vector$, и вычисляет следующий варинат. Также функция build_chain строит цепочку по заданному $r\_vector$ и $p\_vector$. Исходя из всех этих функций был реализован алгоритм дробления вектора. Для последующей проверки гипотезы, введем переменную $default\_m$, о которой будет сказано в следующем пункте.

In [21]:
def next_variant(vector,m,is_p):
    if is_p:
        q=m
    else:
        q=1
    index=m-1
    while index>-1 and vector[index]==1:
        vector[index]=q+index
        index-=1
    if index>-1:
        vector[index]-=1
    return vector,index
def buildChain(indexVector):
    starChain = [1]
    for i in indexVector:
        starChain.append(starChain[-1] + starChain[i - 1])
    return starChain
def div_vector(n,default_m=None):
    if n==1:
        return [1]
    for m in range(lower_estimate(n),high_estimate(n)+1):
        if default_m:
            m=default_m
        q = m // 2
        if (q == 0):
            q = 1
        r_vector=[i for i in range(1,q+1)]
        p_vector=[q+i for i in range(1,m-q+1)]
        index=0
        while index!=-1:
            curChain = buildChain(r_vector + p_vector)
            a_min = curChain[q] + (m - q)
            a_max = curChain[q] * (2 ** (m - q))
            if n<a_min or n>a_max:
                r_vector,index=next_variant(r_vector,q,False)
                continue
            index=m
            while index!=-1:
                # print(r_vector,p_vector)
                curChain=buildChain(r_vector+p_vector)
                if (n == curChain[-1]):
                    return curChain
                p_vector,index=next_variant(p_vector,m-q,True)
            r_vector,index=next_variant(r_vector,q,False)
            p_vector=[i+q for i in range(1,m-q+1)]

Составим таблицу чисел и их минимальных звездных цепочек.

In [31]:
from datetime import datetime
input_array=[1001,1010,1024,1026,1050]
for i in input_array:
    time_now=datetime.now()
    print(div_vector(i))
    print(datetime.now()-time_now)

[1, 2, 4, 8, 16, 32, 64, 128, 192, 200, 400, 800, 1000, 1001]
0:27:02.157982
[1, 2, 4, 8, 16, 32, 64, 128, 256, 320, 336, 672, 1008, 1010]
0:27:28.254791
[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024]
0:00:00.000071
[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1026]
0:00:00.001039
[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1040, 1048, 1050]
0:18:42.732982


| n    | Длина цепочки | Цепочка                                                         | Время работы |
|------|---------------|-----------------------------------------------------------------|--------------|
| 1001 | 14            | [1, 2, 4, 8, 16, 32, 64, 128, 192, 200, 400, 800, 1000, 1001]   | 00:27:02     |
| 1010 | 14            | [1, 2, 4, 8, 16, 32, 64, 128, 256, 320, 336, 672, 1008, 1010]   | 00:27:28     |
| 1024 | 11            | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024]                   | 00:00:01     |
| 1026 | 12            | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1026]             | 00:00:01     |
| 1050 | 14            | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1040, 1048, 1050] | 00:18:42     |

### Вывод.
Алгоритм работает очень долго, однако дает гарантированно минимальную звездную цепочку. Этот алгоритм является переборным, поэтому для больших n является неэффективным.

### Гипотеза Шольца-Брауэра

Для проверки гипотезы, алгоритма доробления вектора, для $2^{12}-1$, поэтому используя соображения о том , что для $n<=64$ выполняется равенство длин.

In [23]:
def Sholz_Brower(n):
    flag=True
    for i in range(1,n+1):
        len_right_equal=len(div_vector(i))+i-1
        if len(div_vector(2**i-1,len_right_equal-1))>len_right_equal:
            print("Гипотеза неверна для")
            print(div_vector(2**i-1),div_vector(i))
            flag=False
            break
        else:
            print("Верно для: ",i)
    if flag:
        print("Гипотеза верна")
Sholz_Brower(12)        

Верно для:  1
Верно для:  2
Верно для:  3
Верно для:  4
Верно для:  5
Верно для:  6
Верно для:  7
Верно для:  8
Верно для:  9
Верно для:  10
Верно для:  11
Верно для:  12
Гипотеза верна


### Вывод.
Было проверена гипотеза Шольца-Брауэра. Для проверки гипотезы понадобилось улучшить алгоритм дробления вектора, так как уже для $n=11$ алгоритм работал непозволительно долго

## Выводы

Были изучены аддитивные цепочки, а также их частный случай - звездные цепочки. Были построены алгоритмы построения этих цепочек, было выяснено, что нет эфффективного алгоритма построения минимальной цепочки, который был бы одновеременно точен и при этом эффективен по времени работы.